### PR TITLE
Add optional Thanos Integration

### PR DESCRIPTION
**Pull request by xrstf**
**What this PR does / why we need it**:
This adds Thanos to our Prometheus chart. I have done this for at least two reasons:

* Our backups keep failing because they take too long. Removing the need to backup Prometheus itself and instead relying on object stores should remove this pain.
* Managing long-term data in Thanos is much easier than in Prometheus.

This PR is meant as a first step to get our feet wet and learn more about the resource requirements.

The Prometheus StatefulSet gets a new sidecar container that ships the blocks to the block storage *and* handles reloading the Prometheus configuration file. We use this to inject the pod name as a `replica` label which is needed for proper deduplication later. This is why the `config-reloader` container is not added when Thanos is enabled.

There could exist a race condition between Prometheus using the templated config from the sidecar before the sidecar has had time to create it. To prevent it from restarting and leaving a "restarts=1" behind in `k get pods`, I put a tiny waiting loop in front of Prometheus. Funny enough, the sidecar expects a running Prometheus before it templates, but it can deal with Prometheus not being ready and simply retries later and refreshes the config again.

In addition to the sidecars we will deploy two redundant Store Gateways in front of our chosen block storage. They will cache frequently used stuff up to the configurable limits.

For actually running Prometheus queries the Query deployment is used. This uses the headless `thanos-store` service to find all the pieces that provide the Store API (i.e. sidecars and store gateways).

First experiments have shown that we can transparently use Thanos instead of Prometheus inside of Grafana, given that we configure the Prometheus sources a bit more flexibly to include Thanos Query's HTTP port (10902).

After feedback at KubeCon I also removed the option to control the TSDB min/max block duration (with Thanos it's explicitly documented to disable compaction in Prometheus and set the min/max to 2h).

In a later PR I will set the resource constraints. Also, the user-cluster Prometheus installations would also need to be updated to include the sidecar.

**Does this PR introduce a user-facing change?**:
```release-note
Added optional Thanos integration to Prometheus for better long-term metrics storage
```

### CODE REVIEW

File: config/monitoring/prometheus/rules/src/general/thanos.yaml [hunk 1]
```diff
@@ -0,0 +1,71 @@
+groups:
```
Comment (kron4eg) [inline config/monitoring/prometheus/rules/src/general/thanos.yaml:1]: shouldn't this somehow take into account that thanos could be disabled?

File: config/monitoring/prometheus/templates/statefulset.yaml [hunk 2]
```diff
@@ -29,25 +34,45 @@ spec:
...
+          set -euo pipefail
+
+          echo "Cleaning up block compaction leftovers..."
+          find /var/prometheus/data -type d -name '*.tmp' -maxdepth 1 -print -exec rm -r {} \;
+
+          {{- if .Values.prometheus.thanos.enabled }}
+          config=/etc/prometheus-shared/prometheus.yaml
+
+          echo "Waiting for $config..."
+          until [ -f $config ]; do sleep 1; done
```
Comment (kron4eg) [inline config/monitoring/prometheus/templates/statefulset.yaml]: This is good opportunity to never see prometheus started :)

Comment (xrstf) [reply config/monitoring/prometheus/templates/statefulset.yaml]: How do you mean?

File: config/monitoring/prometheus/rules/src/general/thanos.yaml (hunk 1 above)
Comment (xrstf) [reply config/monitoring/prometheus/rules/src/general/thanos.yaml:1]: How so? The alerts should only be triggered if the metrics they are based on exist. No Thanos metrics = no Thanos alerts, even if the rules are there. At least that was my idea.

File: config/monitoring/prometheus/templates/statefulset.yaml (hunk 2 above)
Comment (xrstf) [reply config/monitoring/prometheus/templates/statefulset.yaml]: Would it help to `exit 1` after a minute of waiting? Would that improve a broken a situation where Thanos Sidecar cannot start and cannot generate the prometheus.yaml?

Comment (kron4eg) [reply config/monitoring/prometheus/templates/statefulset.yaml]: Some emergency brake wouldn't be excessive.

Comment (xrstf) [reply config/monitoring/prometheus/templates/statefulset.yaml]: Prometheus should now wait for at most one minute and then give up with a more or less helpful error message.

**Review by kron4eg — APPROVED** [review]
/lgtm 
/approve

**Pull request merged**